src/main/java/com/google/devtools/build/lib/remote/http/HttpUploadHandler.java

```diff
@@ -138,8 +143,41 @@ private HttpRequest buildRequest(String path, String host, long contentLength) {
     return request;
   }
 
-  private HttpChunkedInput buildBody(UploadCommand msg) {
-    return new HttpChunkedInput(new ChunkedStream(msg.data()));
+  private HttpChunkedInput buildBody(UploadCommand msg) throws IOException {
+    if (msg.casUpload()) {
+      // we first have to compress to a byte buffer
+      // as we need to know the content length for the header
```

Comment:
This part is extremely ugly, I was hoping we could get away with just streaming the compression to the network, but bazel-remote requires the Content-Length header (even though with chunked input my understanding was you should not have this present). Compressing buffers in memory in Java is super clunky.